src/backend/distributed/connection/connection_management.c

```diff
@@ -437,6 +453,148 @@ ShutdownConnection(MultiConnection *connection)
 }
 
 
+static void
+MultiConnectionReadyExecutePoll(MultiConnectionReady *mcr)
+{
+	MultiConnection *connection = mcr->connection;
+	ConnStatusType status = PQstatus(connection->pgConn);
+
+	Assert(!mcr->ready);
+
+	if (status == CONNECTION_OK)
+	{
+		mcr->ready = true;
+		return;
+	}
+	else if (status == CONNECTION_BAD)
+	{
+		/* FIXME: retries? */
+		mcr->ready = true;
+		return;
+	}
+	else
+	{
+		mcr->ready = false;
+	}
+
+	mcr->pollmode = PQconnectPoll(connection->pgConn);
+
+	/*
+	 * FIXME: Do we want to add transparent retry support here?
+	 */
+	if (mcr->pollmode == PGRES_POLLING_FAILED)
+	{
+		mcr->ready = true;
+		return;
+	}
+	else if (mcr->pollmode == PGRES_POLLING_OK)
+	{
+		mcr->ready = true;
+		return;
+	}
+	else
+	{
+		Assert(mcr->pollmode == PGRES_POLLING_WRITING ||
+			   mcr->pollmode == PGRES_POLLING_READING);
+	}
+}
+
+
+/*
+ *
+ * waitCount populates the number of connections waiting for in case of a non NULL pointer is provided
+ */
+static WaitEventSet *
+WaitEventSetFromMultiConnectionReadyArray(MultiConnectionReady *connections,
+										  int numConnections, int *waitCount)
+{
+	WaitEventSet *waitEventSet = NULL;
+	int connectionIndex = 0;
+
+	/*
+	 * maxEvents is the value min(numConnections, FD_SETSIZE - 3), this is to not add more
+	 * wait events then the set can hold. -3 is used to have room for the following events
+	 * outside of the sockets:
+	 *  - WL_POSTMASTER_DEATH
+	 *  - WL_LATCH_SET
+	 *  - pgwin32_signal_event
+	 *
+	 *  During the loop events are added as long as maxEvents does not reach 0
+	 */
+	int maxEvents = numConnections;
+	if (maxEvents > (FD_SETSIZE - 3))
+	{
+		maxEvents = FD_SETSIZE - 3;
+	}
+
+	if (waitCount)
+	{
+		*waitCount = 0;
+	}
+
+	/* allocate pending connections + 2 for the signal latch and postmaster death */
+	/* (CreateWaitEventSet makes room for pgwin32_signal_event automatically) */
+	waitEventSet = CreateWaitEventSet(CurrentMemoryContext, maxEvents + 2);
```

Comment:
should this be numConnections + 2?

Reply:
There is a limit of events that can be added to the `WaitEventSet` as described in `FD_SETSIZE`.

Most of the time this number will be numConnections as `maxEvents` gets initialized to `numConnections`. However, if the number of connections becomes to big (~1024 on osx) you can't add them all to the `WaitEventSet`.

Instead we keep track of how many connections we can add via `maxEvents`. When this reaches 0 we stop adding them.

Reply:
the naming here is confusing, maxEvents appears to not be the maximum number of events?

Reply:
How do you mean? It is the maximum number of `WaitEvent`s that can be added to the set, hence `maxEvents`. We can make it more verbose by `maxWaitEvents` or `maxWaitEventsToAdd` as it eventually guards the number of times `AddWaitEventToSet` can be called.

Reply:
It might help to introduce some additional variables to make the code more self-documenting, e.g.
- `maxEvents` is the maximum number of events we can add
- `numEventsAdded` is the number of events we added
- `waitEventSetSize = Min(maxEvents, numConnections) + 2` is the size of the wait event set